pkg/core/native_gas_test.go

```diff
@@ -52,3 +58,86 @@ func TestGAS_Roundtrip(t *testing.T) {
+		nKeys     = 4
+	)
+
+	// set Notary nodes and check their balance
+	notaryNodes := make([]*keys.PrivateKey, nNotaries)
+	notaryNodesPublicKeys := make(keys.PublicKeys, nNotaries)
+	for i := range notaryNodes {
+		notaryNodes[i], err = keys.NewPrivateKey()
+		require.NoError(t, err)
+		notaryNodesPublicKeys[i] = notaryNodes[i].PublicKey()
+	}
+	chain.setNodesByRole(t, true, noderoles.P2PNotary, notaryNodesPublicKeys)
+	for _, notaryNode := range notaryNodesPublicKeys {
+		checkBalanceOf(t, chain, notaryNode.GetScriptHash(), 0)
+	}
+	// deposit GAS for `signer` with lock until the next block
+	depositAmount := 100_0000 + (2+int64(nKeys))*transaction.NotaryServiceFeePerKey // sysfee + netfee of the next transaction
+	transferTx := transferTokenFromMultisigAccount(t, chain, notaryHash, gasHash, depositAmount, signer, int64(chain.BlockHeight()+1))
+	res, err := chain.GetAppExecResults(transferTx.Hash(), trigger.Application)
+	require.NoError(t, err)
+	require.Equal(t, vm.HaltState, res[0].VMState)
+	require.Equal(t, 0, len(res[0].Stack))
+
+	// save initial validators balance
+	balances := make(map[int]int64, testchain.ValidatorsCount)
+	for i := 0; i < testchain.ValidatorsCount; i++ {
+		balances[i] = chain.GetUtilityTokenBalance(testchain.PrivateKeyByID(i).GetScriptHash()).Int64()
+	}
+
+	// send transaction with Notary contract as a sender
+	tx := chain.newTestTx(util.Uint160{}, []byte{byte(opcode.PUSH1)})
+	tx.Attributes = append(tx.Attributes, transaction.Attribute{Type: transaction.NotaryAssistedT, Value: &transaction.NotaryAssisted{NKeys: uint8(nKeys)}})
+	tx.NetworkFee = (2 + int64(nKeys)) * transaction.NotaryServiceFeePerKey
+	tx.Signers = []transaction.Signer{
+		{
+			Account: notaryHash,
+			Scopes:  transaction.None,
+		},
+		{
+			Account: signer,
+			Scopes:  transaction.None,
+		},
+	}
+	tx.Scripts = []transaction.Witness{
+		{
+			InvocationScript: append([]byte{byte(opcode.PUSHDATA1), 64}, notaryNodes[0].SignHashable(uint32(testchain.Network()), tx)...),
+		},
+		{
+			InvocationScript:   testchain.Sign(tx),
+			VerificationScript: testchain.MultisigVerificationScript(),
+		},
+	}
+	b := chain.newBlock(tx)
+	require.NoError(t, chain.AddBlock(b))
+	checkBalanceOf(t, chain, notaryHash, int(depositAmount-tx.SystemFee-tx.NetworkFee))
+	singleReward := transaction.NotaryServiceFeePerKey * (nKeys + 1) / nNotaries
+	for _, notaryNode := range notaryNodesPublicKeys {
+		checkBalanceOf(t, chain, notaryNode.GetScriptHash(), singleReward)
+	}
+	for i := 0; i < testchain.ValidatorsCount; i++ {
+		newBalance := chain.GetUtilityTokenBalance(testchain.PrivateKeyByID(i).GetScriptHash()).Int64()
+		expectedBalance := balances[i]
+		if i == int(b.Index)%testchain.CommitteeSize() {
+			// committee reward
+			expectedBalance += 5000_0000
+		}
+		if testchain.IDToOrder(i) == int(b.PrimaryIndex) {
+			// primary reward
+			expectedBalance += tx.NetworkFee - int64(singleReward*nNotaries)
+		}
+		assert.Equal(t, expectedBalance, newBalance, i)
```

Comment:
Maybe it's easier to check for this with `totalSupply` call? GAS claims can interfere, but it's quite predictable otherwise, we add some GAS for committee members and burn system fees.

Reply:
Yes, I've added this check also.